pkg/cloud/api/mutable_resource.go

```diff
@@ -0,0 +1,405 @@
+//
+// If typeTrait is nil, then it will be set to BaseTypeTrait.
+func NewResource[GA any, Alpha any, Beta any](
+	resourceID *cloud.ResourceID,
+	typeTrait TypeTrait[GA, Alpha, Beta],
+) *mutableResource[GA, Alpha, Beta] {
+	if typeTrait == nil {
+		typeTrait = &BaseTypeTrait[GA, Alpha, Beta]{}
+	}
+
+	obj := &mutableResource[GA, Alpha, Beta]{
+		typeTrait:  typeTrait,
+		resourceID: resourceID,
+	}
+
+	// Set .Name from the ResourceID.
+	setName := func(v reflect.Value) {
+		if ft, ok := v.Type().FieldByName("Name"); !ok || ft.Type.Kind() != reflect.String {
+			return
+		}
+		f := v.FieldByName("Name")
+		if !f.IsValid() {
+			panic(fmt.Sprintf("type does not have .Name (%T)", v.Type()))
+		}
+		f.Set(reflect.ValueOf(resourceID.Key.Name))
+	}
+	setName(reflect.ValueOf(&obj.ga).Elem())
+	setName(reflect.ValueOf(&obj.alpha).Elem())
+	setName(reflect.ValueOf(&obj.beta).Elem())
+
+	return obj
+}
+
+// MutableResource wraps the multi-versioned concrete resources.
+type MutableResource[GA any, Alpha any, Beta any] interface {
+	// CheckSchema should be called in init() to ensure that the resource being
+	// wrapped meets the assumptions we are making for this the transformations
+	// to work.
+	CheckSchema() error
+
+	// ResourceID is the resource ID of this resource.
+	ResourceID() *cloud.ResourceID
+
+	// ImpliedVersion returns the best API version for the set of
+	// fields in the resource. It will return an error if it is not
+	// clear which version should be used without missing
+	// configuration.
+	ImpliedVersion() (meta.Version, error)
+
+	// Access the mutable resource.
+	Access(f func(x *GA)) error
+	// AccessAlpha resource.
+	AccessAlpha(f func(x *Alpha)) error
+	// AccessBeta resource.
+	AccessBeta(f func(x *Beta)) error
+
+	// ToGA returns the GA version of this resource. Use error.As
+	// ConversionError to get the specific details.
+	ToGA() (*GA, error)
+	// ToAlpha returns the Alpha version of this resource. Use
+	// error.As ConversionError to get the specific details.
+	ToAlpha() (*Alpha, error)
+	// ToBeta returns the Beta version of this resource. Use
+	// error.As ConversionError to get the specific details.
+	ToBeta() (*Beta, error)
+
+	// Set the value to src. This skips some of the field
+	// validation in Access* so should only be used with a valid
+	// object returned from GCE.
+	Set(src *GA) error
+	// SetAlpha the value to src. This skips some of the field
+	// validation in Access* so should only be used with a valid
+	// object returned from GCE.
+	SetAlpha(src *Alpha) error
+	// SetBeta the value to src. This skips some of the field
+	// validation in Access* so should only be used with a valid
+	// object returned from GCE.
+	SetBeta(src *Beta) error
+
+	// Freeze the resource to a read-only copy. It is an error if it is ambiguous
+	// which version is the correct one i.e. not all fields can be represented in a
+	// single version of the resource.
+	Freeze() (Resource[GA, Alpha, Beta], error)
```

Comment:
Potentially confusing to have a `MutableResource` that you can freeze (and therefore it would no longer be mutable).

Reply:
This is kind of the point. You first are allowed to mutate it, then you Freeze() to remove the possibility of mutation. This is the lifecycle that is desired.